### ISSUE TITLE
Repeat Effect fails to Consolidate

### ISSUE THREAD

**Issue by JousterL** [beat-machine/beat-machine#29]
Stumbled across a bug when using only the Repeat Type, that it won't save because np.concatenate fails with the following:

> ValueError: all the input array dimensions for the concatenation axis must match exactly, but along dimension 1, the array at index 0 has size 2 and the array at index 1 has size 4

It looks like the `return np.tile(beat, shape)` on 158 of periodic.py is scaling the beat out to a size 4. I tried to play around and change it to `np.repeat(beat, self.times, axis=0)` and managed to instead stumble across a possible partial solution for your #14 (every x-th beat is slowed down by the times factor) but while the file will concatenate and save, it's not a doubled beat. I'll admit I'm a bit beyond my kin with my numpy manipulation.

2.1.1 doesn't appear to have the problem, so it looks like a regression in 3.0.0

**Comment by dhsavell** [comment]
Thanks for the report, this seems to be related to #28. Are you able to reproduce this on the latest version, 3.1.1?

**Comment by JousterL** [comment]
Yes, I'm on 3.1.1 currently. I've tried p. much every release on Pypi from 3.0.0 onward and am still encountering this error. I've also done a pip uninstall to make sure the whole package was removed and pulled from 3.1.1

Full call trace below:

> D:\Data\music\T>python -m beatmachine --input "Two Door Cinema Club - Changing Of The Seasons.beat" --output test.mp3 --effects "[{\"type\":\"repeat\", \"period\":2, \"times\": 2}]"
> 
> Applying effect 1/1 (repeat)
> Rendering song
> Traceback (most recent call last):
>   File "C:\Users\JousterL\AppData\Local\Programs\Python\Python38\lib\runpy.py", line 193, in _run_module_as_main
>     return _run_code(code, main_globals, None,
>   File "C:\Users\JousterL\AppData\Local\Programs\Python\Python38\lib\runpy.py", line 86, in _run_code
>     exec(code, run_globals)
>   File "C:\Users\JousterL\AppData\Local\Programs\Python\Python38\lib\site-packages\beatmachine\__main__.py", line 99, in <module>
>     main()
>   File "C:\Users\JousterL\AppData\Local\Programs\Python\Python38\lib\site-packages\beatmachine\__main__.py", line 94, in main
>     beats.save(args.output)
>   File "C:\Users\JousterL\AppData\Local\Programs\Python\Python38\lib\site-packages\beatmachine\beats.py", line 111, in save
>     return self._save_to_file(fp, out_format, extra_ffmpeg_args)
>   File "C:\Users\JousterL\AppData\Local\Programs\Python\Python38\lib\site-packages\beatmachine\beats.py", line 87, in _save_to_file
>     p.communicate(input=self.to_ndarray().tobytes())
>   File "C:\Users\JousterL\AppData\Local\Programs\Python\Python38\lib\site-packages\beatmachine\beats.py", line 53, in to_ndarray
>     return np.concatenate(list(self._beats), axis=0)
>   File "<__array_function__ internals>", line 5, in concatenate
> ValueError: all the input array dimensions for the concatenation axis must match exactly, but along dimension 1, the array at index 0 has size 2 and the array at index 1 has size 4

**Comment by JousterL** [comment]
Huh. So despite pip uninstalling the package to make sure I was grabbing the latest build, it appears I didn't have the change made under #28. I manually updated periodic.py with the change and it's working.

I uninstalled and reinstalled the package from Pypi and the repeat function for 3.1.1 is 
`def process_beat(self, beat: np.ndarray) -> np.ndarray:
        shape = list(np.shape(beat))
        shape[0] = self.times
        return np.tile(beat, shape)`

So it looks like the Pypi package for 3.1.1 doesn't contain #28

**Comment by dhsavell** [comment]
That's weird, I guess I botched the deployment somehow. I'm pushing out a new version 3.1.2 right now that should hopefully contain the latest fixes. Let me know if that works when you get a chance to test it, and thanks for the continued support of the project.

**Comment by JousterL** [comment]
Just pulled it down, same issue. Maybe something is failing with Pypi? I manually pulled both the .tar.gz and the .whl file from https://pypi.org/project/beatmachine/#files and unzipped them, and they've got the old periodic.py code. Version is 3.1.2, update date of today.

**Comment by dhsavell** [comment]
Ah darn, I should stop responding to issues late at night. I forgot to mention that I made a change past #28 specifically for mono audio inputs, which is correctly reflected in 3.1.1 and 3.1.2. So that means there was another bug. I wasn't able to replicate the exception you got but I was getting pretty malformed audio from the effect.

I just fixed this by switching to `np.concatenate` which lets us tile along an axis. 3.1.3 is deploying now since I'd like to have that fix for the webapp ASAP, but I'll leave the issue open until you get a chance to try it.
